src/vanilla/store.ts

```diff
@@ -487,42 +485,58 @@ export const createStore = () => {
       })
       return dependents
     }
-    const loop1 = (a: AnyAtom) => {
-      getDependents(a).forEach((dependent) => {
-        if (dependent !== a) {
-          dependencyMap.set(
-            dependent,
-            (dependencyMap.get(dependent) || new Set()).add(a),
-          )
-          dirtyMap.set(dependent, (dirtyMap.get(dependent) || 0) + 1)
-          loop1(dependent)
+
+    // This is a topological sort via depth-first search, slightly modified from
+    // what's described here for simplicity and performance reasons:
+    // https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+
+    // Step 1: traverse the dependency graph to build the topsorted atom list
+    // We don't bother to check for cycles, which simplifies the algorithm.
+    const topsortedAtoms = new Array<AnyAtom>()
+    const markedAtoms = new Set<AnyAtom>()
+    const visit = (n: AnyAtom) => {
+      if (markedAtoms.has(n)) {
+        return
+      }
+      markedAtoms.add(n)
+      for (const m of getDependents(n)) {
+        if (n !== m) {
```

Comment:
should this be `!isSelfAtom(n, m)`?

Reply:
Yeah, probably. (Or `!isSelfAtom(m, n)`?)
But, jotai-scope seems to be working for now. cc @yf-yang
We need tests for `unstable_is`...

Reply:
For what it’s worth, this line was brought from the previous implementation and it seems to be the a pretty common check in jotai (though I’m not sure why it’s necessary at all, seems like the set of dependents should never contain self, so we shouldn’t need to check for it. But it’s harmless enough.)

Reply:
> this line was brought from the previous implementation

Yes, it's a leftover in #2356.

> seems like the set of dependents should never contain self

good point. can we simply remove the if statement?
(So, dependencies can contain self, but dependents can't, right? I mean that's true AFAIR.)

Reply:
If dependencies can contain self, shouldn’t the opposite be true?

It didn’t seem important to remove it if unnecessary. So seems safer to just leave it there for now. I wouldn’t want proving whether we need or don’t need it to hold up this improvement.

Reply:
I would like to remove it if possible. Let me check the code.

Reply:
That was my misunderstanding. Dependents can contain self.

Let's do this for now.
```suggestion
        if (!isSelfAtom(n, m)) {
```

Reply:
Intuitively it should be `!==`, not `isSelfAtom`, because the scoped atom copy could depend on the original atom internally. Empirically, I suggest we simply give it a try with all the jotai-scope examples. Just give me a commit id and I can try it.

Reply:
😂 latest commit 69b5557a does break jotai-scope, yes something needs to change.

Reply:
yeah, I was skeptical about it. okay, let's revert.

Reply:
@dai-shi [34a635a](https://github.com/pmndrs/jotai/pull/2363/commits/34a635a159045a4398163ab43e7203fffeb158a1) should be reverted, because [c991752](https://github.com/pmndrs/jotai/pull/2363/commits/c991752654bf382ec3d14e1fd85dd6cb5b048f9b) is the correct version.

Reply:
> Intuitively it should be `!==`, not `isSelfAtom`, because the scoped atom copy could depend on the original atom internally. Empirically, I suggest we simply give it a try with all the jotai-scope examples. Just give me a commit id and I can try it.

I don't think this is true anymore. Scope uses the original atom as the key, so any scoped reference to that atom will always return the scoped version, even if its a self-reference.

Reply:
sounds good.